### PR TITLE
Automatically reconnect when the database connection drops

### PR DESCRIPTION
**Pull request by hmac**
If/when the database connection drops, catch the error and attempt to
reconnect, backing off exponentially.

### CODE REVIEW

File: lib/que_ui/app.rb [hunk 1]
```diff
@@ -4,9 +4,14 @@
 
 require "que_ui/queries"
 require "que_ui/query_builder"
+require "que_ui/database"
 
 module QueUI
   class App < Sinatra::Base
+    configure do
+      Database.pg_connect!
```
Comment (petehamilton) [inline lib/que_ui/app.rb]: `connect!`?

Comment (hmac) [inline lib/que_ui/app.rb]: yeah!?

File: lib/que_ui/database.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+          user: ENV["PGUSER"],
+          password: ENV["PGPASSWORD"],
+          dbname: ENV["PGDATABASE"],
+        )
+      end
+
+      # Attempt to reconnect to the database, backing off exponentially
+      def attempt_reconnect!
+        @next_attempt ||= Time.now
+        @disconnected_at ||= Time.now
```
Comment (hmac) [inline lib/que_ui/database.rb]: Should reverse L16 and L17 so `@disconnected_at` is an earlier time than `@next_attempt`

File: lib/que_ui/database.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        @next_attempt = Time.now + delay
+      end
+
+      private
+
+      # The delay is calculated as the time between the most recent disconnection
+      # and the next scheduled reconnection, multiplied by 2.
+      # This will simulate an exponential backoff, capped at 1 minute.
+      def delay
+        [(@next_attempt - (@disconnected_at)) * 2, 60].min
```
Comment (hmac) [inline lib/que_ui/database.rb]: Redundant parens

Comment (hmac) [inline lib/que_ui/database.rb]: Consider a 1 second minimum delay?

**Comment by hmac** [comment]
@petehamilton made some changes, should be a bit easier to read now

File: lib/que_ui/database.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+        connect!
+        reset!
+      rescue PG::ConnectionBad
+        @next_attempt = Time.now + delay
+      end
+
+      private
+
+      # The delay will double each time it's applied
```
Comment (petehamilton) [inline lib/que_ui/database.rb]: Not sure this comment is needed, code is super clear I think

File: lib/que_ui/database.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          user: ENV["PGUSER"],
+          password: ENV["PGPASSWORD"],
+          dbname: ENV["PGDATABASE"],
+        )
+      end
+
+      # Attempt to reconnect to the database, backing off exponentially
+      def attempt_reconnect!
+        @next_attempt ||= Time.now
+        return unless @next_attempt < Time.now
```
Comment (petehamilton) [inline lib/que_ui/database.rb]: Could remove `@next_attempt ||= Time.now` and do `@next_attempt.nil? || @next_attempt < Time.now`?

Comment (hmac) [inline lib/que_ui/database.rb]: Considered that, but I thought that would be less clear. Not so sure now, to be honest..

File: lib/que_ui/database.rb (hunk 4 above)
Comment (hmac) [inline lib/que_ui/database.rb]: 👍 ✂️

**Comment by petehamilton** [comment]
LGTM, nice work 💯

**Pull request merged**